core/src/main/java/com/linecorp/armeria/server/healthcheck/HealthCheckService.java

```diff
@@ -0,0 +1,420 @@
+                final boolean currentHealthiness = isHealthy();
+                if (isHealthy == currentHealthiness) {
+                    final CompletableFuture<HttpResponse> future = new CompletableFuture<>();
+                    final ScheduledFuture<Boolean> timeoutFuture = ctx.eventLoop().schedule(
+                            () -> future.complete(HttpResponse.of(notModifiedHeaders)),
+                            longPollingTimeoutMillis, TimeUnit.MILLISECONDS);
+                    final PendingResponse pendingResponse = new PendingResponse(method, future, timeoutFuture);
+                    if (isHealthy) {
+                        pendingUnhealthyResponses.add(pendingResponse);
+                    } else {
+                        pendingHealthyResponses.add(pendingResponse);
+                    }
+
+                    updateRequestTimeout(ctx, longPollingTimeoutMillis);
+                    return HttpResponse.from(future);
+                } else {
+                    // State has been changed before we acquire the lock.
+                    // Fall through because there's no need for long polling.
+                }
+            }
+        }
+
+        switch (method) {
+            case HEAD:
+            case GET:
+                return newResponse(method, isHealthy);
+            case CONNECT:
+            case DELETE:
+            case OPTIONS:
+            case TRACE:
+                return HttpResponse.of(HttpStatus.METHOD_NOT_ALLOWED);
+        }
+
+        if (updateHandler == null) {
+            return HttpResponse.of(HttpStatus.METHOD_NOT_ALLOWED);
+        }
+
+        return HttpResponse.from(updateHandler.handle(ctx, req).thenApply(updateResult -> {
+            if (updateResult != null) {
+                serverHealth.setHealthy(updateResult);
+            }
+            return HttpResponse.of(newResponse(method, isHealthy()));
+        }));
+    }
+
+    private boolean isHealthy() {
+        for (HealthChecker healthChecker : healthCheckers) {
+            if (!healthChecker.isHealthy()) {
+                return false;
+            }
+        }
+        return serverHealth.isHealthy();
+    }
+
+    private long getLongPollingTimeoutMillis(HttpRequest req) {
+        if (!isLongPollingEnabled()) {
+            return 0;
+        }
+
+        final String prefer = req.headers().get(HttpHeaderNames.PREFER);
+        if (prefer == null) {
+            return 0;
+        }
+
+        // TODO(trustin): Optimize this once https://github.com/line/armeria/issues/1835 is resolved.
+        final LongHolder timeoutMillisHolder = new LongHolder();
+        try {
+            ArmeriaHttpUtil.parseDirectives(prefer, (name, value) -> {
+                if ("wait".equals(name)) {
+                    timeoutMillisHolder.value = TimeUnit.SECONDS.toMillis(Long.parseLong(value));
+                }
+            });
+        } catch (NumberFormatException ignored) {
+            // Malformed "wait" value.
```

Comment:
I think explicitly throwing the bad request exception here is easier to follow the logic.

Reply:
Done